### DISCUSSION TITLE
Need Some Guidance in BranchPythonOperator

### DISCUSSION THREAD

**Discussion by Nidesh-Arumugam** [apache/airflow#19517] [Q&A]
I need to do a bash command at first, followed by this based on the logs i need to do another set of bash commands.

Detailed Explanation

For example: i am doing a bash task to check the status of a device. Once the task is finished based on the logs, If the status is online i need to perform another  bash task, if the status is offline i need to perform alternative   bash task.

can any one give some idea or guidance, it would be a great help :)
Thanks

**Comment by josh-fell** [comment]
How are you running, or plan to run, the bash commands?

**Reply by Nidesh-Arumugam** [reply]
planning to run them in a single Dag with a list of tasks. First task will be a bash task(b1) using bash operator, next based on the log of the previous task(b1) i need to run few  tasks. If b1 is success need to run b2, else b3 needs to be executed.

**Comment by josh-fell** [comment]
When using the `BashOperator` the last line that is written to stdout will be pushed as an `XCom`. So you could run your bash command to check the log for whatever status you need and make sure the status is the last entry written to stdout. Then, this `XCom` can be used as an input to the `python_callable` set in the `BranchPythonOperator`. The `python_callable` can then return the `task_id` of the branch you want the pipeline to follow.

Here is a very simplified example. The first task generates a random number between 1 and 10. The number is written to stdout and therefore pushed as an `XCom`. This output is used as the input arg to the `_check_log_output()` function to determine which branch to follow: "b2" if even, "b3" if odd.
```python
from datetime import datetime, timedelta

from airflow.decorators import dag
from airflow.operators.bash import BashOperator
from airflow.operators.python import BranchPythonOperator


def _check_log_output(result):
    return "b2" if int(result) % 2 else "b3"


@dag(schedule_interval=None, start_date=datetime(2021, 11, 6))
def bash_branch():
    b1 = BashOperator(task_id="b1", bash_command="echo $((1 + $RANDOM % 10))")

    branch = BranchPythonOperator(task_id="branch", python_callable=_check_log_output, op_args=[b1.output])

    b2 = BashOperator(task_id="b2", bash_command="echo Followed branch to task b2")

    b3 = BashOperator(task_id="b3", bash_command="echo Followed branch to task b3")

    b1 >> branch >> [b2, b3]


dag = bash_branch()
```

**Reply by Nidesh-Arumugam** [reply]
thx for the idea, The above example is working good. But what's the case if i am getting multiple lines from the logs. Instead of fetching the last line and branching them based on condition, i need to determine the branching based on the line at the  top or at the middle?

In other words, what's the practice to push the entire log to Xcom instead of last line ?
